zcash_client_sqlite/src/wallet.rs

```diff
@@ -808,213 +808,449 @@ pub(crate) fn get_derived_account<P: consensus::Parameters>(
     accounts.next().transpose()
 }
 
+#[derive(Debug)]
+pub(crate) struct Progress {
+    scan: Option<Ratio<u64>>,
+    recover: Option<Ratio<u64>>,
+}
+
 pub(crate) trait ScanProgress {
-    fn sapling_scan_progress(
+    fn sapling_scan_progress<P: consensus::Parameters>(
         &self,
         conn: &rusqlite::Connection,
+        params: &P,
         birthday_height: BlockHeight,
+        recover_until_height: Option<BlockHeight>,
         fully_scanned_height: BlockHeight,
         chain_tip_height: BlockHeight,
-    ) -> Result<Option<Ratio<u64>>, SqliteClientError>;
+    ) -> Result<Progress, SqliteClientError>;
 
     #[cfg(feature = "orchard")]
-    fn orchard_scan_progress(
+    fn orchard_scan_progress<P: consensus::Parameters>(
         &self,
         conn: &rusqlite::Connection,
+        params: &P,
         birthday_height: BlockHeight,
+        recover_until_height: Option<BlockHeight>,
         fully_scanned_height: BlockHeight,
         chain_tip_height: BlockHeight,
-    ) -> Result<Option<Ratio<u64>>, SqliteClientError>;
+    ) -> Result<Progress, SqliteClientError>;
 }
 
 #[derive(Debug)]
 pub(crate) struct SubtreeScanProgress;
 
-impl ScanProgress for SubtreeScanProgress {
-    #[tracing::instrument(skip(conn))]
-    fn sapling_scan_progress(
-        &self,
-        conn: &rusqlite::Connection,
-        birthday_height: BlockHeight,
-        fully_scanned_height: BlockHeight,
-        chain_tip_height: BlockHeight,
-    ) -> Result<Option<Ratio<u64>>, SqliteClientError> {
-        if fully_scanned_height == chain_tip_height {
-            // Compute the total blocks scanned since the wallet birthday
-            conn.query_row(
-                "SELECT SUM(sapling_output_count)
-                 FROM blocks
-                 WHERE height >= :birthday_height",
+#[allow(clippy::too_many_arguments)]
+fn subtree_scan_progress<P: consensus::Parameters>(
+    conn: &rusqlite::Connection,
+    params: &P,
+    table_prefix: &'static str,
+    output_count_col: &'static str,
+    shard_height: u8,
+    pool_activation_height: BlockHeight,
+    birthday_height: BlockHeight,
+    recover_until_height: Option<BlockHeight>,
+    fully_scanned_height: BlockHeight,
+    chain_tip_height: BlockHeight,
+) -> Result<Progress, SqliteClientError> {
+    let mut stmt_scanned_count_between = conn.prepare_cached(&format!(
+        "SELECT SUM({output_count_col})
+        FROM blocks
+        WHERE :start_height <= height AND height <= :end_height",
+    ))?;
+    let mut stmt_scanned_count_from = conn.prepare_cached(&format!(
+        "SELECT SUM({output_count_col})
+        FROM blocks
+        WHERE :start_height <= height",
+    ))?;
+    let mut stmt_start_tree_size = conn.prepare_cached(&format!(
+        "SELECT MAX({table_prefix}_commitment_tree_size - {output_count_col})
+        FROM blocks
+        WHERE height <= :start_height",
```

Comment:
Is it intended that `height <= :start_height` overlaps `:start_height <= height` (in the queries above) by one block?

Reply:
Yes, because subtree boundaries don't fall at block boundaries.